From 3496717a9bb8f5500556fa407a11b0fd785de72c Mon Sep 17 00:00:00 2001
From: Dave Airlie <airlied@redhat.com>
Date: Fri, 24 Apr 2026 18:50:36 +0300
Subject: [PATCH 1/2] c11/threads: fix build on c23

C23/glibc is now including once_init in stdlib.h

https://patchwork.sourceware.org/project/glibc/patch/78061085-f04a-0c45-107b-5a8a15521083@redhat.com/#213088

Just fix up our use of it.

Cc: mesa-stable
Reviewed-by: Yonggang Luo <luoyonggang@gmail.com>
(cherry picked from commit 179e744f7577d98df7c79d7324c22acfb32a0154)

Backported onto mesa 22.0.2: the original patch touched the post-refactor
files src/c11/threads.h and src/c11/impl/threads_posix.c; in 22.0.2 the
equivalent definitions still live together in include/c11/threads_posix.h,
so the same __once_flag_defined guards are applied there.

Drop this patch on the next mesa uprev.

[Backported to 22.0.2]
Signed-off-by: Adrian Ratiu <adrian.ratiu@collabora.com>
---
 include/c11/threads_posix.h | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/include/c11/threads_posix.h b/include/c11/threads_posix.h
index 802526a77c8..beeed34d709 100644
--- a/include/c11/threads_posix.h
+++ b/include/c11/threads_posix.h
@@ -51,7 +51,9 @@ Configuration macro:
 #include <pthread.h>
 
 /*---------------------------- macros ----------------------------*/
+#ifndef __once_flag_defined
 #define ONCE_FLAG_INIT PTHREAD_ONCE_INIT
+#endif
 #ifdef INIT_ONCE_STATIC_INIT
 #define TSS_DTOR_ITERATIONS PTHREAD_DESTRUCTOR_ITERATIONS
 #else
@@ -66,7 +68,9 @@ typedef pthread_cond_t  cnd_t;
 typedef pthread_t       thrd_t;
 typedef pthread_key_t   tss_t;
 typedef pthread_mutex_t mtx_t;
+#ifndef __once_flag_defined
 typedef pthread_once_t  once_flag;
+#endif
 
 
 /*
@@ -90,11 +94,13 @@ impl_thrd_routine(void *p)
 
 /*--------------- 7.25.2 Initialization functions ---------------*/
 // 7.25.2.1
+#ifndef __once_flag_defined
 static inline void
 call_once(once_flag *flag, void (*func)(void))
 {
     pthread_once(flag, func);
 }
+#endif
 
 
 /*------------- 7.25.3 Condition variable functions -------------*/
-- 
2.53.0

